### PR TITLE
Change type of constant sized arrays to `NonZeroU32`

### PR DESCRIPTION
**Pull request by teoxoy**
Split from https://github.com/gfx-rs/naga/pull/2266

### CODE REVIEW

File: tests/wgsl-errors.rs [hunk 1]
```diff
@@ -921,26 +921,52 @@ fn invalid_arrays() {
...
+    check(
+        "alias Bad = array<f32, -1>;",
+        r###"error: array element count must resolve to an integer scalar (u32, i32)
+  ┌─ wgsl:1:24
+  │
+1 │ alias Bad = array<f32, -1>;
+  │                        ^^ must resolve to u32/i32
+
+"###,
+    );
```
Comment (teoxoy) [inline tests/wgsl-errors.rs:969]: Note that this will be correct (saying that it needs to be positive as the one above) with a follow-up.

**Comment by jimblandy** [comment]
@teoxoy How will we support arrays whose lengths are override constants? Will that be a third variant of `ArraySize`?

**Comment by teoxoy** [comment]
That's the plan.

**Review by jimblandy — APPROVED** [review]
This looks good. A few suggestions:

File: src/proc/mod.rs [hunk 2]
```diff
@@ -199,7 +199,7 @@ impl super::TypeInner {
     }
 
     /// Get the size of this type.
-    pub fn size(&self, gctx: GlobalCtx) -> u32 {
+    pub fn size(&self, _gctx: GlobalCtx) -> u32 {
```
[the remark this answers is not in the document]
Comment (teoxoy) [reply src/proc/mod.rs:202]: I'm not sure how exactly we'll deal with overrides yet, so I didn't want to take it out just to add it back in later.

**Pull request merged**